cpp/dolfinx/la/MatrixCSR.h

```diff
@@ -0,0 +1,491 @@
+  {
+    const std::int32_t local_size0 = _index_maps[0]->size_local();
+    const std::int32_t num_ghosts0 = _index_maps[0]->num_ghosts();
+
+    // For each ghost row, pack and send values to send to neighborhood
+    std::vector<int> insert_pos(_val_send_disp);
+    std::vector<T> ghost_value_data(_val_send_disp.back());
+    for (int i = 0; i < num_ghosts0; ++i)
+    {
+      const int neighbor_rank = _ghost_row_to_neighbor_rank[i];
+
+      // Get position in send buffer
+      const std::int32_t val_pos = insert_pos[neighbor_rank];
+      std::copy(std::next(_data.data(), _row_ptr[local_size0 + i]),
+                std::next(_data.data(), _row_ptr[local_size0 + i + 1]),
+                std::next(ghost_value_data.begin(), val_pos));
+      insert_pos[neighbor_rank]
+          += _row_ptr[local_size0 + i + 1] - _row_ptr[local_size0 + i];
+    }
+
+    _ghost_value_data_in.resize(_val_recv_disp.back());
+
+    // Compute data sizes for send and receive from displacements
+    std::vector<int> val_send_count(_val_send_disp.size() - 1);
+    std::adjacent_difference(std::next(_val_send_disp.begin()),
+                             _val_send_disp.end(), val_send_count.begin());
+
+    std::vector<int> val_recv_count(_val_recv_disp.size() - 1);
+    std::adjacent_difference(std::next(_val_recv_disp.begin()),
+                             _val_recv_disp.end(), val_recv_count.begin());
+
+    int status = MPI_Ineighbor_alltoallv(
+        ghost_value_data.data(), val_send_count.data(), _val_send_disp.data(),
+        dolfinx::MPI::mpi_type<T>(), _ghost_value_data_in.data(),
+        val_recv_count.data(), _val_recv_disp.data(),
+        dolfinx::MPI::mpi_type<T>(), _comm.comm(), &_request);
+    assert(status == MPI_SUCCESS);
+  }
+
+  /// Begin communication of ghost row data to owning ranks
+  /// @note Must be preceded by MatrixCSR::finalize_begin()
+  void finalize_end()
+  {
+    int status = MPI_Wait(&_request, MPI_STATUS_IGNORE);
+    assert(status == MPI_SUCCESS);
+
+    // Add to local rows
+    assert(_ghost_value_data_in.size() == _unpack_pos.size());
+    for (std::size_t i = 0; i < _ghost_value_data_in.size(); ++i)
+      _data[_unpack_pos[i]] += _ghost_value_data_in[i];
+
+    // Set ghost row data to zero
+    const std::int32_t local_size0 = _index_maps[0]->size_local();
+    std::fill(std::next(_data.begin(), _row_ptr[local_size0]), _data.end(), 0);
+  }
+
+  /// Index maps for the row and column space. The row IndexMap contains
+  /// ghost entries for rows which may be inserted into and the column
+  /// IndexMap contains all local and ghost columns that may exist in
+  /// the owned rows.
+  ///
+  /// @return Row (0) and column (1) index maps
+  const std::array<std::shared_ptr<const common::IndexMap>, 2>&
+  index_maps() const
+  {
+    return _index_maps;
+  }
+
+  /// Get local data values
+  /// @param[in] ghost_rows Set to true to include data of ghost rows
+  xtl::span<T> values(bool ghost_rows = false)
```

Comment:
This is a low-level function so I think we can let the caller deal with ghosts by having a function that returns he pointer to the owned values.

Best to avoid default arguments whenever reasonably possible.